reddit.py

```diff
@@ -0,0 +1,141 @@
+"""Reddit source code and datastore model classes.
+
+"""
+import logging
+
+from granary import reddit as gr_reddit
+from granary import source as gr_source
+from oauth_dropins import reddit as oauth_reddit
+from oauth_dropins.webutil.util import json_dumps, json_loads
+import webapp2
+from webob import exc
+
+import models
+import util
+
+
+class Reddit(models.Source):
+  """A Reddit account.
+
+  The key name is the username.
+  """
+  GR_CLASS = gr_reddit.Reddit
+  OAUTH_START_HANDLER = oauth_reddit.StartHandler
+  SHORT_NAME = 'reddit'
+  TYPE_LABELS = {
+    'post': 'submission',
+    'comment': 'comment',
+  }
+  TRANSIENT_ERROR_HTTP_CODES = ('404',)
+
+  @staticmethod
+  def new(handler, auth_entity=None, **kwargs):
+    """Creates and returns a :class:`Twitter` entity.
+
+    Args:
+      handler: the current :class:`webapp2.RequestHandler`
+      auth_entity: :class:`oauth_dropins.twitter.TwitterAuth`
+      kwargs: property values
+    """
+    user = json_loads(auth_entity.user_json)
+    gr_source = gr_reddit.Reddit(auth_entity.refresh_token)
+    return Reddit(id=user.get('name'),
+                  auth_entity=auth_entity.key,
+                  url=gr_source.user_url(user.get('name')),
+                  name=user.get('name'),
+                  picture=user.get('icon_img'),
+                  **kwargs)
+
+  def silo_url(self):
+    """Returns the Twitter account URL, e.g. https://twitter.com/foo."""
+    return self.gr_source.user_url(self.key_id())
+
+  def label_name(self):
+    """Returns the username."""
+    return self.key_id()
+
+  def search_for_links(self):
+    """Searches for activities with links to any of this source's web sites.
+
+    Twitter search supports OR:
```

Comment:
remove these two paragraphs? feel free to update with reddit-specific details if necessary, otherwise omit